nbsite/pyodide/__init__.py

```diff
@@ -243,69 +284,105 @@ def terminate(cls, *args):
         """
         Terminates a running process.
         """
-        if not cls._current_process:
-            return
-        try:
-            cls._conn.send({'type': 'close'})
-            cls._current_process.join()
-        except Exception:
-            cls._current_process.terminate()
-        finally:
-            cls._current_process = None
+        for state in cls._state.values():
+            if state['process'] is None:
+                return
+            try:
+                state['conn'].send({'type': 'close'})
+                state['process'].join()
+            except Exception:
+                state['process'].terminate()
+        cls._state.clear()
 
     @classmethod
-    def _launch_process(cls, timeout=5):
+    def _launch_process(cls, source, timeout=5):
         """
         Launches a process to execute code in.
         """
-        cls.terminate()
-        cls._conn, child_conn = Pipe()
-        cls._current_process = Process(target=cls._execution_process, args=(child_conn,))
-        cls._current_process.start()
+        cls._state[source]['conn'], child_conn = Pipe()
+        cls._state[source]['process'] = process = Process(
+            target=cls._execution_process, args=(child_conn,), daemon=True
+        )
+        process.start()
 
     @classmethod
-    def _kill(cls):
-        cls._current_process.terminate()
-        cls._current_process = None
+    def _kill(cls, source):
+        if cls._state[source]['process']:
+            cls._state[source]['process'].terminate()
+        del cls._state[source]
 
     def run(self):
+        outdir = self.state.document.settings.env.app.outdir
+        cache_path = pathlib.Path(str(outdir)) / '.pyodide'
+        cache_path.mkdir(exist_ok=True)
         current_source = self.state_machine.get_source()
-        if self._current_source != current_source or self._current_process is None:
-            PyodideDirective._current_count = 0
-            PyodideDirective._current_source = current_source
-            self._launch_process()
+        if current_source not in self._state:
+            with open(current_source, encoding='utf-8') as src_doc:
+                content = src_doc.read()
+            content_hash = hashlib.md5(content.encode('utf-8')).hexdigest()
+            cache_file = cache_path / f'{content_hash}.json'
+            if cache_file.is_file():
+                with open(cache_file, encoding='utf-8') as cf:
+                    self._state[current_source]['cache'] = json.load(cf)
+            else:
+                self._launch_process(current_source)
+            total = content.count('{pyodide}')
+            self._state[current_source]['total'] = total
+            self._state[current_source]['hash'] = content_hash
+        else:
+            self._state[current_source]['count'] += 1
+        state = self._state[current_source]
+        count = state['count']
+        cache_file = cache_path / f'{state["hash"]}.json'
```

Comment:
Could the cache accidentally be uploaded with the rest of the docs build? 

Maybe use `platformdirs.user_cache_path()` for this?

Reply:
Pretty certain dotted directories don't get uploaded. This is the correct place to cache since we want to be able to reset the cache by clearing the build directory, also Sphinx itself uses `<build-dir>/.doctrees` as a cache for its internal data structures.